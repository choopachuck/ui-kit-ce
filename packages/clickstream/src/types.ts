export type InputElementType = JSX.IntrinsicElements['input']['type']

export type BaseElementType = keyof JSX.IntrinsicElements

export type ButtonElementType = JSX.IntrinsicElements['button']['type']

export type LinkAnchorElementType = 'link' | 'anchor'

export type ClickStreamElementProperty = { key: string; value: string }

export type ClickStreamEventBatch<
  TData extends unknown = ClickStreamEventData
> = ClickStreamEventBatchItem<TData>[]

export type ClickStreamElementType =
  | InputElementType
  | BaseElementType
  | LinkAnchorElementType
  | ButtonElementType

export type ClickStreamBaseData = {
  /**
   * URL страницы, где было инициировано событие
   */
  url: string
  /**
   * Дата и время, в которое было инициировано событие в формате `ISO`
   */
  timeStamp: string
  /**
   * Название страницы, где было инициировано событие
   */
  pageName: string
  /**
   * Тип соединения пользователя с интернетом на момент инициирования события
   */
  connectionType: string
  /**
   * Уровень заряда устройства пользователя на момент инициирования события
   */
  batteryLevel: string
}

export type ClickStreamEventData = ClickStreamBaseData & {
  /**
   * Массив с атрибутами HTML-элемента, с которым было инициировано событие
   */
  elementAttributes: ClickStreamElementProperty[]

  /**
   * Тип инициированного события. По умолчанию принимает значение события `event.type`.
   * Также может принимать любое пользовательское значение
   */
  elementEvent: string
  /**
   * Тип элемента, с которым произошло событие. Если событие произошло с элементом `input`,
   * то значением будет являться `element.type`, иначе принимает значение, равное
   * `element.tagName` в нижнем регистре
   */
  elementType: ClickStreamElementType
  /**
   * HTML-тэг элемента, с которым произошло событие
   */
  elementTag: string
  /**
   * Значение элемента, с которым произошло событие. Если событие произошло с элементом `input`
   * и типом `radio` или `checkbox`, то значением будет являться `element.checked`. Если
   * событие произошло с элементом `input` и типом не `radio` или `checkbox`, то значением
   * будет являться `element.value`. Иначе принимает значение, равное `element.innerText`
   */
  value: string
}

export type ClickStreamEventMeta = {
  /**
   * Дата и время, в которое была отправлена очередь событиий в в формате `ISO`
   */
  timeStamp: string
  /**
   * Константное значение, которое означает, что очередь событий была отправлена через веб-приложение
   */
  platform: 'WEB'
  /**
   * Браузер, в котором была отправлена очередь событий
   */
  browser: string
  /**
   * Установленный языковой пакет в браузере пользователя на момент отправки очереди событий
   */
  systemLanguage: string
  /**
   * Язык, установленный в самом веб-приложении на момент отправки очереди событий
   */
  applicationLanguage: string
  /**
   * Размер экрана устройства пользователя на момент отправки очереди событий
   */
  screenSize: string
}

export type ClickStreamContextProps = {
  /**
   * Функция для вызова события
   *
   * @param {Event} event Инициированное событие
   * @param {string} eventType Тип инициированного события. По умолчанию принимает значение события
   * `event.type`. Также может принимать любое пользовательское значение
   * @param {boolean} immediate Флаг отправки события без очереди
   */
  dispatchEvent: (event: Event, eventType?: string, immediate?: boolean) => void
}

export type ClickStreamProviderProps<
  TData extends unknown = ClickStreamEventData
> = {
  /**
   * Функция обратного вызова, которая срабатывает при первой инициализации провайдера
   *
   * @param {ClickStreamBaseData<TData>} data Данные первой инициализации провайдера
   *
   * @param {ClickStreamEventMeta} meta Метаданные отправляемой очереди событий
   */
  onInit?: (data: ClickStreamBaseData, meta: ClickStreamEventMeta) => void
  /**
   * Функция обратного вызова, которая срабатывает при отправке очереди событий
   *
   * После отправки очереди событий, текущая очередь очищается
   *
   * @param {ClickStreamEventBatch<TData>} batch Отправляемая очередь событий, размер которой равен
   * значению свойства `batchSize`
   *
   * @param {ClickStreamEventMeta} meta Метаданные отправляемой очереди событий
   */
  onSendEvent?: (
    batch: ClickStreamEventBatch<TData>,
    meta: ClickStreamEventMeta
  ) => void
  /**
   * Функция для получения отформатированных данных для событий `ClickStreamProvider`
   *
   * @param {Event} event Инициированное событие
   * @param {string} eventType Тип инициированного события. По умолчанию принимает значение события
   * `event.type`. Также может принимать любое пользовательское значение
   */
  formatEventData?: (event: Event, eventType: string) => TData
  /**
   * Функция обратного вызова, которая срабатывает при добавлении нового события в очередь событий
   *
   * @param {Event} event Инициированное событие
   * @param {TData} data Отформатированные данные для события
   * @param {ClickStreamEventBatch<TData>} batch Накопленная очередь событий
   * @param {boolean} immediate Флаг отправки события без очереди
   */
  onBatch?: (
    event: Event,
    data: TData,
    batch: ClickStreamEventBatch<TData>,
    immediate: boolean
  ) => void
  /**
   * Флаг для отключения у всего приложения автоматической отправки событий через `ClickStreamProvider`
   */
  disableAutoSendEvent?: boolean
  /**
   * Максимальный размер очереди событий
   */
  batchSize?: number
  /**
   * Разрешенное время бездействия пользователя в миллисекундах
   *
   * Если пользователь совершил количество действий меньше, чем указано в свойстве `batchSize`, то
   * по истечению указанного времени `ClickStreamProvider` автоматически запустит функцию обратного
   * вызова `onSendEvent` с накопившейся очередью событий
   */
  inactivityTime?: number
}

export type ClickStreamEventBatchItem<
  TData extends unknown = ClickStreamEventData
> = {
  /**
   * Инициированное событие
   */
  event: Event
  /**
   * Отформатированные данные для события
   */
  data: TData
}
