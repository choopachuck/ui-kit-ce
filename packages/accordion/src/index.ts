export * from './Accordion'
export * from './AccordionItem'
export type {
  AccordionIconProps,
  AccordionComponentsConfig,
} from './components'
