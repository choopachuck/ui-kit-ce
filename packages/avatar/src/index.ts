export * from './Avatar'
export * from './AvatarGroup'
export * from './types'
export type {
  AvatarGroupExtraProps,
  AvatarGroupComponentsConfig,
} from './components'
export * from './classes'
export * from './constants'
