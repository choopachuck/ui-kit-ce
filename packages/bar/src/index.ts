export * from './Bar'
export * from './BarButton'
export * from './BarDate'
export * from './BarDivider'
export * from './BarDropdown'
export * from './BarDropdownItem'
export * from './BarItem'
export * from './BarMenuItem'
export * from './BarSearch'
export * from './BarSelect'
export * from './constants'
export * from './interfaces'
export * from './SubBar'
