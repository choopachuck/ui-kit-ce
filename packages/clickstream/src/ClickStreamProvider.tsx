'use client'

/* eslint-disable react/prop-types */ // TODO: косячит eslint при типизации пропсов компонента через дженерики React.forwardRef

import * as React from 'react'
import {
  getElementAttributes,
  getElementValue,
  getTagName,
  getApplicationLanguage,
  getBatteryLevel,
  getBrowser,
  getConnectionType,
  getPageName,
  getScreenSize,
  getSystemLanguage,
  getISOTime,
  getClickEventElement,
  getChangeEventElement,
  isForbiddenForElement,
} from './utils'
import {
  ClickStreamElementType,
  ClickStreamEventData,
  ClickStreamEventMeta,
  ClickStreamContextProps,
  ClickStreamProviderProps,
  ClickStreamEventBatch,
  ClickStreamEventBatchItem,
} from './types'
import {
  CLICK_STREAM_EVENTS,
  CLICK_STREAM_BATCH_SIZE,
  CLICK_STREAM_INACTIVITY_TIME,
} from './constants'

const EVENT_TYPE_TO_HANDLER_MAP: Record<
  string,
  (event: Event) => [ClickStreamElementType, HTMLElement]
> = {
  click: getClickEventElement,
  change: getChangeEventElement,
} as const

const generateBaseData = async () => ({
  url: window.location.href,
  batteryLevel: await getBatteryLevel(),
  connectionType: getConnectionType(),
  pageName: getPageName(),
  timeStamp: getISOTime(),
})

const makeProcessData =
  (elementType: ClickStreamElementType, element: HTMLElement) =>
  async (_: Event, elementEvent: string): Promise<ClickStreamEventData> => {
    const baseData = await generateBaseData()

    return {
      elementEvent,
      value: String(getElementValue(element)),
      elementType,
      elementTag: getTagName(element),
      elementAttributes: getElementAttributes(element),
      ...baseData,
    }
  }

const generateMeta = (): ClickStreamEventMeta => ({
  applicationLanguage: getApplicationLanguage(),
  browser: getBrowser(),
  platform: 'WEB',
  screenSize: getScreenSize(),
  systemLanguage: getSystemLanguage(),
  timeStamp: getISOTime(),
})

const processDispatchToBody = (
  dispatch: (
    event: Event,
    elementEvent?: string,
    immediate?: boolean
  ) => Promise<void>,
  strategy: 'add' | 'remove'
) => {
  const _dispatch = dispatch as ClickStreamContextProps['dispatchEvent']

  CLICK_STREAM_EVENTS.forEach((event) => {
    if (strategy === 'add') {
      document.body.addEventListener(event, _dispatch)
    }
    if (strategy === 'remove') {
      document.body.removeEventListener(event, _dispatch)
    }
  })
}

export type ClickStreamRef = Pick<ClickStreamContextProps, 'dispatchEvent'>

const ClickStreamContext = React.createContext({} as ClickStreamContextProps)

const _ClickStreamProvider = React.forwardRef<
  ClickStreamRef,
  React.PropsWithChildren<ClickStreamProviderProps>
>(
  (
    {
      onSendEvent,
      formatEventData: formatEventDataProp,
      disableAutoSendEvent,
      children,
      batchSize = CLICK_STREAM_BATCH_SIZE,
      onBatch,
      inactivityTime = CLICK_STREAM_INACTIVITY_TIME,
      onInit,
    },
    ref
  ): JSX.Element => {
    const batchRef =
      React.useRef<ClickStreamEventBatch<ClickStreamEventData> | null>([])

    const dispatchTimeout = React.useRef<ReturnType<typeof setTimeout>>()

    const debouncedDispatchEvent = React.useCallback(
      async (
        event: Event,
        elementEvent?: string,
        immediate = false,
        verify = true
      ) => {
        if (dispatchTimeout.current) {
          clearTimeout(dispatchTimeout.current)
        }

        if (!batchRef.current) {
          return
        }

        const [elementType, element] = (
          EVENT_TYPE_TO_HANDLER_MAP[event.type] || getClickEventElement
        )(event)

        if (verify && isForbiddenForElement(element, event.type)) {
          return
        }

        const formatEventData =
          formatEventDataProp || makeProcessData(elementType, element)

        const data = await Promise.resolve(
          formatEventData(event, elementEvent ?? event.type)
        )
        const newEventData: ClickStreamEventBatchItem = { event, data }
        const meta = generateMeta()

        onBatch?.(event, data, batchRef.current)

        // Если очередь событий достигла допустимого размера или событие отправлено с флагом `immediate`,
        // то сразу отправляем очередь пользователю, иначе оборачиваем отправку в отложенный вызов на
        // случай, если пользователь не достигнет допустипого размера очереди по истечению разрешенного
        // времени бездействия

        const releaseBatch = () => {
          if (!batchRef.current || !batchRef.current.length) {
            return
          }

          onSendEvent?.(batchRef.current, meta)
          batchRef.current = []
        }

        // Отложенный вызов делаем только в том случае, если `inactivityTime` является
        // неотрицательным числом. Специальные значения, такие как `NaN`, `Infinity`
        // отключают отложенный вызов
        if (Number.isFinite(inactivityTime) && Math.sign(inactivityTime) >= 0) {
          dispatchTimeout.current = setTimeout(releaseBatch, inactivityTime)
        }

        if (immediate) {
          onSendEvent?.([newEventData], meta)

          return
        }

        batchRef.current.push(newEventData)

        if (batchSize === batchRef.current.length) {
          releaseBatch()
        }
      },
      [formatEventDataProp, onSendEvent, batchSize, onBatch, inactivityTime]
    )

    const value = React.useMemo<ClickStreamContextProps>(
      () => ({
        dispatchEvent: (
          event: Event,
          elementEvent?: string,
          immediate?: boolean
        ) => debouncedDispatchEvent(event, elementEvent, immediate, false),
      }),
      [debouncedDispatchEvent]
    )

    React.useEffect(() => {
      if (disableAutoSendEvent) {
        return
      }
      const dispatchClickStreamEvent = async (event: Event) => {
        await debouncedDispatchEvent(event, event.type)
      }

      processDispatchToBody(dispatchClickStreamEvent, 'add')

      return () => {
        processDispatchToBody(dispatchClickStreamEvent, 'remove')
      }
    }, [debouncedDispatchEvent, disableAutoSendEvent])

    React.useImperativeHandle(
      ref,
      () => ({
        dispatchEvent: value.dispatchEvent,
      }),
      [value.dispatchEvent]
    )

    React.useEffect(() => {
      //eslint-disable-next-line no-void
      void (async () => {
        const baseData = await generateBaseData()

        onInit?.(baseData, generateMeta())
      })()
    }, [onInit])

    return (
      <ClickStreamContext.Provider value={value}>
        {children}
      </ClickStreamContext.Provider>
    )
  }
)

export const ClickStreamProvider = _ClickStreamProvider as <
  TData extends unknown = ClickStreamEventData
>(
  props: React.PropsWithChildren<ClickStreamProviderProps<TData>> & {
    ref?: React.ForwardedRef<ClickStreamRef>
  },
  ref: React.ForwardedRef<ClickStreamRef>
) => JSX.Element

export const useClickStreamContext = (): ClickStreamContextProps => {
  const clickStreamContext = React.useContext(ClickStreamContext)

  if (!clickStreamContext) {
    throw new Error(
      'ClickStream context must be in ClickStream context provider!'
    )
  }

  return clickStreamContext
}
