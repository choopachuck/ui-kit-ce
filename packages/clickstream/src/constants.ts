export const CLICK_STREAM_EVENTS = ['click', 'change'] as const

export const CLICK_STREAM_INACTIVITY_TIME = 30000

export const CLICK_STREAM_BATCH_SIZE = 10

// Нотация нарушена для обратной совместимости с @sbol/clickstream-agent
export const CONNECTION_TYPES = {
  bluetooth: 'BLUETOOTH',
  cellular: 'CELLULAR',
  ethernet: 'ETHERNET',
  wifi: 'WIFI',
  mixed: 'UNKNOWN',
  none: 'UNKNOWN',
  wimax: 'UNKNOWN',
  other: 'UNKNOWN',
  unknown: 'UNKNOWN',
} as const

export const CLICK_STREAM_GENERIC_ATTRIBUTES = {
  OFF: 'data-click-stream-off',
} as const

export const ALIGNMENT = 10

export const MINUTES_IN_HOUR = 60

export const MS_IN_SECOND = 1000

export const SECONDS_IN_MINUTE = 60

export const MS_IN_MINUTE = SECONDS_IN_MINUTE * MS_IN_SECOND
