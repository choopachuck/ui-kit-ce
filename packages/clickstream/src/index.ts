export * from './ClickStreamProvider'
export type {
  ClickStreamBaseData,
  ClickStreamContextProps,
  ClickStreamElementProperty,
  ClickStreamElementType,
  ClickStreamEventBatch,
  ClickStreamEventBatchItem,
  ClickStreamEventData,
  ClickStreamEventMeta,
  ClickStreamProviderProps,
} from './types'
