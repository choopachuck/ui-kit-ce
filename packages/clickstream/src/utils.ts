import {
  ClickStreamElementType,
  BaseElementType,
  InputElementType,
  ClickStreamElementProperty,
} from './types'
import {
  CONNECTION_TYPES,
  CLICK_STREAM_GENERIC_ATTRIBUTES,
  ALIGNMENT,
  MINUTES_IN_HOUR,
  MS_IN_MINUTE,
} from './constants'

declare global {
  interface Navigator {
    getBattery: <T>() => Promise<Record<string, T>>
    browserLanguage: string
    userLanguage: string
    systemLanguage: string
    connection?: {
      type?: keyof typeof CONNECTION_TYPES
    }
  }
}

export const isForbiddenForElement = (
  element: HTMLElement,
  eventType: string
): boolean => {
  if (
    element.hasAttribute(CLICK_STREAM_GENERIC_ATTRIBUTES.OFF) ||
    getTagName(element) === 'body'
  ) {
    return true
  }

  if (eventType === 'click') {
    if (
      element.hasAttribute('aria-disabled') &&
      element.getAttribute('aria-disabled') === 'true'
    ) {
      return true
    }

    if (
      isInput(element) ||
      isImg(element) ||
      element.hasAttribute('role') ||
      isSvg(element) ||
      isButton(element)
    ) {
      return false
    }

    return !element.innerText
  }

  return false
}

export const getClickEventElement = (
  event: Event
): [ClickStreamElementType, HTMLElement] => {
  const eventsPath = event.composedPath()
  const fallbackElements: Partial<
    Record<BaseElementType, [ClickStreamElementType, HTMLElement]>
  > = {}
  const defaultElement = [
    getTagName(eventsPath[0] as HTMLElement),
    eventsPath[0],
  ] as [ClickStreamElementType, HTMLElement]

  for (let i = 0; i <= eventsPath.length; i++) {
    const element = eventsPath[i] as HTMLElement
    if (isLabel(element)) {
      return [getTagName(element), element]
    }
    if (isInput(element)) {
      const input = element as HTMLInputElement

      return [input.type as ClickStreamElementType, input]
    }
    if (isButton(element)) {
      return [
        (element as HTMLButtonElement).type as ClickStreamElementType,
        element,
      ]
    }
    if (element instanceof HTMLLinkElement) {
      return ['link', element]
    }
    if (element instanceof HTMLAnchorElement) {
      return ['anchor', element]
    }
    if (element instanceof HTMLElement && element.hasAttribute('role')) {
      return [getTagName(element), element]
    }

    if (isSvg(element)) {
      fallbackElements.svg = [getTagName(element), element]
    }
  }

  return fallbackElements.svg || defaultElement
}

export const getChangeEventElement = (
  event: Event
): [ClickStreamElementType, HTMLInputElement | HTMLTextAreaElement] => {
  const element = event.target as HTMLInputElement | HTMLTextAreaElement
  const tagName = getTagName(element)

  if (tagName === 'select') {
    return [tagName as ClickStreamElementType, element]
  }

  if (element instanceof HTMLTextAreaElement) {
    return ['textarea', element]
  }

  return [element.type as InputElementType, element]
}

export const getElementAttributes = (
  element: HTMLElement
): ClickStreamElementProperty[] => {
  const attributes: ClickStreamElementProperty[] = []
  for (let i = 0, atts = element.attributes, n = atts.length; i < n; i++) {
    attributes.push({ key: atts[i].nodeName, value: String(atts[i].nodeValue) })
  }

  return attributes
}

export const isInput = (element: HTMLElement): boolean =>
  element instanceof HTMLInputElement || element instanceof HTMLTextAreaElement

export const isImg = (element: HTMLElement): boolean =>
  element instanceof HTMLImageElement

export const isSvg = (element: HTMLElement): boolean =>
  element instanceof SVGElement

export const isButton = (element: HTMLElement): boolean =>
  element instanceof HTMLButtonElement

export const isCheckboxOrRadio = (element: HTMLInputElement): boolean =>
  ['checkbox', 'radio'].includes(element.type)

export const isLabel = (element: HTMLElement): boolean =>
  element instanceof HTMLLabelElement

export const getTagName = (element: HTMLElement): BaseElementType =>
  element?.tagName?.toLowerCase() as BaseElementType

export const getElementValue = (
  element: HTMLElement
): string | boolean | null => {
  if (isInput(element)) {
    const input = element as HTMLInputElement
    const value = isCheckboxOrRadio(input) ? input.checked : input.value

    return value
  }

  if (isImg(element)) {
    return element.getAttribute('src')
  }

  return (
    element.getAttribute('aria-label') ||
    element.innerText ||
    element.textContent ||
    ''
  )
}

export const getApplicationLanguage = (): string =>
  globalThis.document?.documentElement.lang.substring(0, 16) || 'ru'

export const getBatteryLevel = async (): Promise<string> => {
  try {
    // Поддерживается не всеми браузерами
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-assignment
    const battery = (await globalThis.navigator?.getBattery<number>()) || {}

    const { level } = battery

    if (typeof level === 'number') {
      return String(100 * level)
    }

    return '0'
  } catch (error) {
    return '0'
  }
}

export const getBrowser = (): string => globalThis.navigator?.userAgent

export const getConnectionType = (): string =>
  CONNECTION_TYPES[
    (globalThis.navigator?.connection?.type as keyof typeof CONNECTION_TYPES) ??
      'unknown'
  ]

export const getPageName = (): string => {
  if (!globalThis.document) {
    return ''
  }

  return globalThis.document.title
}

export const getUrl = (): string => {
  if (!window?.location?.href) {
    return ''
  }

  return window.location.href
}

export const getScreenSize = (): string =>
  globalThis.screen
    ? `${globalThis.screen.width}x${globalThis.screen.height}`
    : '0x0'

export const getSystemLanguage = (): string =>
  globalThis.navigator?.language ||
  globalThis.navigator?.userLanguage ||
  globalThis.navigator?.browserLanguage ||
  globalThis.navigator?.systemLanguage ||
  'ru'

const padStart = (value: number): number | string => {
  return value < ALIGNMENT ? `0${value}` : value
}

const formatOffset = (offset: number): string => {
  return `${offset < 0 ? '+' : '-'}${padStart(
    Math.abs(Math.floor(offset / MINUTES_IN_HOUR))
  )}:${padStart(Math.abs(offset % MINUTES_IN_HOUR))}`
}

export const getISOTime = (): string => {
  const date = new Date()
  const correctedDate = new Date(
    date.valueOf() - date.getTimezoneOffset() * MS_IN_MINUTE
  )
  const [correctedShortDate] = correctedDate.toISOString().split('Z')

  return `${correctedShortDate}${formatOffset(
    correctedDate.getTimezoneOffset()
  )}`
}
