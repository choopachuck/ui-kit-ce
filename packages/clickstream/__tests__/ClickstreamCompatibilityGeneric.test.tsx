/*--------------------------------------------------------------------------------------------------
 * --------- Тестирование совместимости нативных HTML-элементов с провайдером ClickStream ----------
 * -------------------------------------------------------------------------------------------------
 */

import * as React from 'react'
import { render, waitFor } from '@testing-library/react'
import * as ClickStreamUtils from '../src/utils'
import userEvent from '@testing-library/user-event'
import { ClickStreamProvider } from '../src'
import { mockGlobals } from './helpers'

jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01'))
jest
  .spyOn(ClickStreamUtils, 'getISOTime')
  .mockImplementation(() => '2020-01-01T00:00:00.078-00:00')

beforeAll(() => {
  mockGlobals()
})

it('clickstream lib components compatibility: native HTML-elements', async () => {
  const handleBatch = jest.fn()
  const handleSendEvent = jest.fn()
  const expectedBatchCalls = 13

  const { getByText, getByTitle } = render(
    <ClickStreamProvider
      batchSize={expectedBatchCalls}
      onSendEvent={handleSendEvent}
      onBatch={handleBatch}
    >
      <button data-id="1" id="custom-button" aria-label="custom button">
        button-01
      </button>
      <button className="custom-class">button-02</button>
      <button>
        <span>button-03</span>
      </button>
      <div>
        <button style={{ color: 'red' }}>button-04</button>
      </div>
      <input
        title="input-01"
        aria-label="3"
        type="text"
        defaultValue="some value"
      />
      <div title="empty-div">
        <span>
          <p></p>
        </span>
      </div>
      {/** role="button" - необходимо указать интерактивную роль, чтобы отработал event bubbling в тесте */}
      <header role="button">header</header>
      <textarea aria-label="my awesome textarea" defaultValue="textarea" />
      <input title="text-input" type="text" placeholder="Placeholder text..." />
      <input title="radio-input" type="radio" data-foo="bar" />
      <input title="checkbox-input" type="checkbox" />
    </ClickStreamProvider>
  )
  await waitFor(() => userEvent.click(getByText('button-01')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByText('button-02')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByText('button-03')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByText('button-04')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByTitle('input-01')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByTitle('empty-div')), { timeout: 1 }) // событие не происходит, так как кликстрим не ловит события с пустых тэгов
  await waitFor(() => userEvent.click(getByText('header')), { timeout: 1 })
  await waitFor(() => userEvent.click(getByText('textarea')), { timeout: 1 })

  await waitFor(() => userEvent.click(getByTitle('text-input')), { timeout: 1 })
  userEvent.keyboard('hello world')
  await waitFor(() => userEvent.tab(), { timeout: 1 })

  await waitFor(() => userEvent.click(getByTitle('radio-input')), {
    timeout: 1,
  })
  await waitFor(() => userEvent.click(getByTitle('checkbox-input')), {
    timeout: 1,
  })

  await waitFor(() =>
    expect(handleBatch).toHaveBeenCalledTimes(expectedBatchCalls)
  )
  expect(handleSendEvent).toMatchSnapshot()
})
