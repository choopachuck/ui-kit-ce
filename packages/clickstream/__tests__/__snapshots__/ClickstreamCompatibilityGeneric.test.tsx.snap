// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clickstream lib components compatibility: native HTML-elements 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "data-id",
                "value": "1",
              },
              Object {
                "key": "id",
                "value": "custom-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "class",
                "value": "custom-class",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button-02",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button-03",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "style",
                "value": "color: red;",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button-04",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "input-01",
              },
              Object {
                "key": "aria-label",
                "value": "3",
              },
              Object {
                "key": "type",
                "value": "text",
              },
              Object {
                "key": "value",
                "value": "some value",
              },
            ],
            "elementEvent": "click",
            "elementTag": "input",
            "elementType": "text",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "some value",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "role",
                "value": "button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "header",
            "elementType": "header",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "header",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "aria-label",
                "value": "my awesome textarea",
              },
            ],
            "elementEvent": "click",
            "elementTag": "textarea",
            "elementType": "textarea",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "textarea",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "text-input",
              },
              Object {
                "key": "type",
                "value": "text",
              },
              Object {
                "key": "placeholder",
                "value": "Placeholder text...",
              },
            ],
            "elementEvent": "click",
            "elementTag": "input",
            "elementType": "text",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "text-input",
              },
              Object {
                "key": "type",
                "value": "text",
              },
              Object {
                "key": "placeholder",
                "value": "Placeholder text...",
              },
            ],
            "elementEvent": "change",
            "elementTag": "input",
            "elementType": "text",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "hello world",
          },
          "event": Event {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "radio-input",
              },
              Object {
                "key": "type",
                "value": "radio",
              },
              Object {
                "key": "data-foo",
                "value": "bar",
              },
            ],
            "elementEvent": "click",
            "elementTag": "input",
            "elementType": "radio",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "true",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "radio-input",
              },
              Object {
                "key": "type",
                "value": "radio",
              },
              Object {
                "key": "data-foo",
                "value": "bar",
              },
            ],
            "elementEvent": "change",
            "elementTag": "input",
            "elementType": "radio",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "true",
          },
          "event": Event {
            "isTrusted": true,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "checkbox-input",
              },
              Object {
                "key": "type",
                "value": "checkbox",
              },
            ],
            "elementEvent": "click",
            "elementTag": "input",
            "elementType": "checkbox",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "true",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "title",
                "value": "checkbox-input",
              },
              Object {
                "key": "type",
                "value": "checkbox",
              },
            ],
            "elementEvent": "change",
            "elementTag": "input",
            "elementType": "checkbox",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "true",
          },
          "event": Event {
            "isTrusted": true,
          },
        },
      ],
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
